src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -843,23 +1077,49 @@ typedef struct client_channel_call_data {
   grpc_error *error;
 
   grpc_lb_policy *lb_policy;  // Holds ref while LB pick is pending.
-  grpc_closure lb_pick_closure;
-  grpc_closure lb_pick_cancel_closure;
+  grpc_closure pick_closure;
+  grpc_closure pick_cancel_closure;
 
   grpc_connected_subchannel *connected_subchannel;
   grpc_call_context_element subchannel_call_context[GRPC_CONTEXT_COUNT];
   grpc_polling_entity *pollent;
-
-  grpc_transport_stream_op_batch *waiting_for_pick_batches[MAX_WAITING_BATCHES];
-  size_t waiting_for_pick_batches_count;
-  grpc_closure handle_pending_batch_in_call_combiner[MAX_WAITING_BATCHES];
-
-  grpc_transport_stream_op_batch *initial_metadata_batch;
-
   grpc_linked_mdelem lb_token_mdelem;
 
-  grpc_closure on_complete;
-  grpc_closure *original_on_complete;
+  // Batches are added to this list when received from above.
+  // They are removed when we are done handling the batch (i.e., when
+  // either we have invoked all of the batch's callbacks or we have
+  // passed the batch down to the subchannel call and are not
+  // intercepting any of its callbacks).
+  pending_batch pending_batches[MAX_PENDING_BATCHES];
+  bool pending_send_initial_metadata : 1;
+  bool pending_send_message : 1;
+  bool pending_send_trailing_metadata : 1;
+
+  // Retry state.
+  bool enable_retries : 1;
+  bool retry_committed : 1;
+  bool last_attempt_got_server_pushback : 1;
+  int num_attempts_completed;
+  size_t bytes_buffered_for_retry;
+  grpc_backoff retry_backoff;
+  grpc_timer retry_timer;
+
+  // Cached data for retrying send ops.
+  // send_initial_metadata
+  bool seen_send_initial_metadata;
+  grpc_linked_mdelem *send_initial_metadata_storage;
+  grpc_metadata_batch send_initial_metadata;
+  uint32_t send_initial_metadata_flags;
+  gpr_atm *peer_string;
+  // send_message
+  // When we get a send_message op, we replace the original byte stream
+  // with a grpc_caching_byte_stream that caches the slices to a
+  // local buffer for use in retries.
+  grpc_core::InlinedVector<grpc_byte_stream_cache *, 3> send_messages;
```

Comment:
Was this inlined size chosen for performance reasons or picked from a hat? If the latter, add a todo to point this out as a magic number tuning opportunity.